Fix conntrack entry removal on endpoint leave and agent restart

tgraf opened the pull request:
So far, no scrubbing was performed and all entries in the conntrack table have
been subject to regular expiration cycles. When starting the agent, the agent
restores all previous endpoints. The IPs of all those endpoints are known. We
can thus scrub the conntrack table from all entries that cannot be associated
with one of the restored endpoints.

Testing:

Old table
```
$ sudo cilium bpf ct list global
ICMPv6 IN [f00d::a0f:0:0:fbac]:0 -> [fd01::b]:0 related  expires=98121 rx_packets=1 rx_bytes=142 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
TCP OUT [2a00:1450:400a:801::2004]:80 -> [f00d::a0f:0:0:fbac]:43232  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=0 revnat=0 src_sec_id=63927
ICMPv6 OUT [2a00:1450:400a:801::2004]:0 -> [f00d::a0f:0:0:fbac]:0 related  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=10 revnat=0 src_sec_id=63927

TCP OUT 172.217.168.68:80 -> 10.11.248.28:41732  expires=119587 rx_packets=10 rx_bytes=13101 tx_packets=11 tx_bytes=692 flags=10 revnat=0 src_sec_id=63927
TCP OUT 216.58.215.228:80 -> 10.11.248.28:58144  expires=119661 rx_packets=13 rx_bytes=13303 tx_packets=14 tx_bytes=854 flags=10 revnat=0 src_sec_id=63927
ICMP OUT 8.8.8.8:0 -> 10.11.248.28:0 related  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=74 flags=10 revnat=0 src_sec_id=63927
ICMP OUT 216.58.215.228:0 -> 10.11.248.28:0 related  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=74 flags=10 revnat=0 src_sec_id=63927
UDP OUT 8.8.8.8:53 -> 10.11.248.28:52753  expires=98121 rx_packets=1 rx_bytes=102 tx_packets=1 tx_bytes=74 flags=0 revnat=0 src_sec_id=63927
UDP OUT 8.8.8.8:53 -> 10.11.248.28:52597  expires=98121 rx_packets=1 rx_bytes=90 tx_packets=1 tx_bytes=74 flags=0 revnat=0 src_sec_id=63927
```

[restart agent]

New table:
```
sudo cilium bpf ct list global
```

When an endpoint is removed, all conntrack entries of that endpoint are
guaranteed to become invalid. Use the opportunity to remove all such entries.

Testing:
```
$ sudo cilium bpf ct list global
ICMPv6 IN [f00d::a0f:0:0:3adc]:0 -> [fd01::b]:0 related  expires=100021 rx_packets=1 rx_bytes=142 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
ICMPv6 OUT [2a00:1450:400a:803::200e]:0 -> [f00d::a0f:0:0:3adc]:0 related  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=10 revnat=0 src_sec_id=1031
TCP OUT [2a00:1450:400a:803::200e]:80 -> [f00d::a0f:0:0:3adc]:37958  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=0 revnat=0 src_sec_id=1031

UDP OUT 8.8.8.8:53 -> 10.11.246.180:48671  expires=100021 rx_packets=1 rx_bytes=98 tx_packets=1 tx_bytes=70 flags=0 revnat=0 src_sec_id=1031
ICMP IN 10.11.224.225:0 -> 10.11.0.1:0 related  expires=100022 rx_packets=1 rx_bytes=74 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
TCP IN 10.11.224.225:4240 -> 10.11.0.1:54450  expires=121562 rx_packets=5 rx_bytes=475 tx_packets=5 tx_bytes=464 flags=10 revnat=0 src_sec_id=1
ICMP OUT 8.8.8.8:0 -> 10.11.246.180:0 related  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=70 flags=10 revnat=0 src_sec_id=1031
UDP OUT 8.8.8.8:53 -> 10.11.246.180:40772  expires=100021 rx_packets=1 rx_bytes=86 tx_packets=1 tx_bytes=70 flags=0 revnat=0 src_sec_id=1031
ICMP OUT 172.217.168.78:0 -> 10.11.246.180:0 related  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=74 flags=10 revnat=0 src_sec_id=1031
TCP OUT 172.217.168.78:80 -> 10.11.246.180:36166  expires=121561 rx_packets=5 rx_bytes=814 tx_packets=6 tx_bytes=418 flags=10 revnat=0 src_sec_id=1031

$ docker rm -f 98a611fb2d4f

$ sudo cilium bpf ct list global
ICMP IN 10.11.224.225:0 -> 10.11.0.1:0 related  expires=100022 rx_packets=1 rx_bytes=74 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
TCP IN 10.11.224.225:4240 -> 10.11.0.1:54450  expires=121562 rx_packets=5 rx_bytes=475 tx_packets=5 tx_bytes=464 flags=10 revnat=0 src_sec_id=1
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5243)
<!-- Reviewable:end -->

tgraf committed:
conntrack: Scrub all entries that cannot be associated with restored endpoints

So far, no scrubbing was performed and all entries in the conntrack table have
been subject to regular expiration cycles. When starting the agent, the agent
restores all previous endpoints. The IPs of all those endpoints are known. We
can thus scrub the conntrack table from all entries that cannot be associated
with one of the restored endpoints.

Testing:

Old table
```
$ sudo cilium bpf ct list global
ICMPv6 IN [f00d::a0f:0:0:fbac]:0 -> [fd01::b]:0 related  expires=98121 rx_packets=1 rx_bytes=142 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
TCP OUT [2a00:1450:400a:801::2004]:80 -> [f00d::a0f:0:0:fbac]:43232  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=0 revnat=0 src_sec_id=63927
ICMPv6 OUT [2a00:1450:400a:801::2004]:0 -> [f00d::a0f:0:0:fbac]:0 related  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=10 revnat=0 src_sec_id=63927

TCP OUT 172.217.168.68:80 -> 10.11.248.28:41732  expires=119587 rx_packets=10 rx_bytes=13101 tx_packets=11 tx_bytes=692 flags=10 revnat=0 src_sec_id=63927
TCP OUT 216.58.215.228:80 -> 10.11.248.28:58144  expires=119661 rx_packets=13 rx_bytes=13303 tx_packets=14 tx_bytes=854 flags=10 revnat=0 src_sec_id=63927
ICMP OUT 8.8.8.8:0 -> 10.11.248.28:0 related  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=74 flags=10 revnat=0 src_sec_id=63927
ICMP OUT 216.58.215.228:0 -> 10.11.248.28:0 related  expires=98121 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=74 flags=10 revnat=0 src_sec_id=63927
UDP OUT 8.8.8.8:53 -> 10.11.248.28:52753  expires=98121 rx_packets=1 rx_bytes=102 tx_packets=1 tx_bytes=74 flags=0 revnat=0 src_sec_id=63927
UDP OUT 8.8.8.8:53 -> 10.11.248.28:52597  expires=98121 rx_packets=1 rx_bytes=90 tx_packets=1 tx_bytes=74 flags=0 revnat=0 src_sec_id=63927
```

[restart agent]

New table:
```
sudo cilium bpf ct list global
```

Related: #5239

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
conntrack: Remove GCFilterType

Required refactoring to allow simple addition of removal of certain IPs while
performing GC.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Remove conntrack entries on endpoint removal

When an endpoint is removed, all conntrack entries of that endpoint are
guaranteed to become invalid. Use the opportunity to remove all such entries.

Testing:
```
$ sudo cilium bpf ct list global
ICMPv6 IN [f00d::a0f:0:0:3adc]:0 -> [fd01::b]:0 related  expires=100021 rx_packets=1 rx_bytes=142 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
ICMPv6 OUT [2a00:1450:400a:803::200e]:0 -> [f00d::a0f:0:0:3adc]:0 related  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=10 revnat=0 src_sec_id=1031
TCP OUT [2a00:1450:400a:803::200e]:80 -> [f00d::a0f:0:0:3adc]:37958  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=94 flags=0 revnat=0 src_sec_id=1031

UDP OUT 8.8.8.8:53 -> 10.11.246.180:48671  expires=100021 rx_packets=1 rx_bytes=98 tx_packets=1 tx_bytes=70 flags=0 revnat=0 src_sec_id=1031
ICMP IN 10.11.224.225:0 -> 10.11.0.1:0 related  expires=100022 rx_packets=1 rx_bytes=74 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
TCP IN 10.11.224.225:4240 -> 10.11.0.1:54450  expires=121562 rx_packets=5 rx_bytes=475 tx_packets=5 tx_bytes=464 flags=10 revnat=0 src_sec_id=1
ICMP OUT 8.8.8.8:0 -> 10.11.246.180:0 related  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=70 flags=10 revnat=0 src_sec_id=1031
UDP OUT 8.8.8.8:53 -> 10.11.246.180:40772  expires=100021 rx_packets=1 rx_bytes=86 tx_packets=1 tx_bytes=70 flags=0 revnat=0 src_sec_id=1031
ICMP OUT 172.217.168.78:0 -> 10.11.246.180:0 related  expires=100021 rx_packets=0 rx_bytes=0 tx_packets=1 tx_bytes=74 flags=10 revnat=0 src_sec_id=1031
TCP OUT 172.217.168.78:80 -> 10.11.246.180:36166  expires=121561 rx_packets=5 rx_bytes=814 tx_packets=6 tx_bytes=418 flags=10 revnat=0 src_sec_id=1031

$ docker rm -f 98a611fb2d4f

$ sudo cilium bpf ct list global
ICMP IN 10.11.224.225:0 -> 10.11.0.1:0 related  expires=100022 rx_packets=1 rx_bytes=74 tx_packets=0 tx_bytes=0 flags=10 revnat=0 src_sec_id=1
TCP IN 10.11.224.225:4240 -> 10.11.0.1:54450  expires=121562 rx_packets=5 rx_bytes=475 tx_packets=5 tx_bytes=464 flags=10 revnat=0 src_sec_id=1
```

Fixes: #5239

Signed-off-by: Thomas Graf <thomas@cilium.io>

jrfastab reviewed (approved):
Looks good to me.

aanm commented on pkg/endpointmanager/conntrack.go:
```diff
@@ -82,12 +106,13 @@ func EnableConntrackGC(ipv4, ipv6 bool, gcinterval int) {
+			if len(eps) > 0 || initialScan {
 				if ipv6 {
-					runGC(nil, true, ctmap.NewGCFilterBy(ctmap.GCFilterByTime))
+					runGC(nil, true, createGCFilter(false, initialScan, restoredEndpoints))
```
`s/createGCFilter(false/createGCFilter(true`

tgraf committed:
endpoint: Clear conntrack on initial endpoint build

On regular endpoint lifecycle management, the connection tracking table should
automatically be cleared from entries of endpoints that have been removed.

Similar, if the agent crashed, the cleaning during the bootstrapping should
remove all entries that cannot be related to a restored endpoint.

As a last safety net, this commit removes all eventual connection tracking
entries related to an endpoint IP that is being re-used. The cleaning is done
for the first regeneration of the endpoint in parallel to the compilation.

Fixes: #5239

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -539,6 +542,15 @@ func (f *GCFilter) doFiltering(srcIP net.IP, dstIP net.IP, dstPort uint16, nextH
+		_, srcIPExists := f.MatchIPs[srcIP.String()]
+		_, dstIPExists := f.MatchIPs[dstIP.String()]
+		if srcIPExists || dstIPExists {
+			log.Infof("Removing entry due to match %v %v", srcIPExists, dstIPExists)
```
Should we move this to Debug? Those messages can be verbosed for a long-lived endpoint with long-lived connections.

Also, replace `%v` with `%t` and write a better message, `Removing entry due to match true true` doesn't seem useful

tgraf replied:
I've removed it completely. I've used it to verify correctness.
